report-ng/app/src/components/method-details/method.ts

```diff
@@ -173,14 +176,34 @@ export class Method {
             });
 
             if (!routeConfig.hasChildRouter) {
-                const navOptions = {replace:true};
+                const navOptions = {replace: true};
                 const enabledRouteConfig = this._router.routes.find(routeConfig => routeConfig.nav);
                 this._router.navigateToRoute(enabledRouteConfig.name, {}, navOptions);
             }
         });
+
+        // Get all method details
+        this._statistics.getExecutionStatistics().then(executionStatistic => {
+            executionStatistic.classStatistics.forEach(classStatistic => {
+                classStatistic.methodContexts
+                    .filter(methodContext => methodContext.methodType == MethodType.TEST_METHOD)
+                    .forEach(methodContext => this._allMethodDetails.push(new MethodDetails(methodContext, classStatistic)));
+            });
+            // Sort for run index
+            this._allMethodDetails = this._allMethodDetails.sort((a, b) => a.methodContext.methodRunIndex - b.methodContext.methodRunIndex);
+        });
```

Comment:
Looks quite heavy to me to sort all methods, keep them cached and search them every time in the ViewModel to show the previous and next methods.

Please keep in mind, that functions cannot be observed in Aurelia. Thats why Aurelia is calling functions every 10ms. So your method `_getMethodDetailsFromCurrent()` gets called very often!

My lighweight proposal would be something like this:
```typescript
this._statistics.getExecutionStatistics().then(executionStatistics => {
    const myRunIndex = this._methodDetails.methodContext.methodRunIndex
    executionStatistics.classStatistics
        .flatMap(value => value.methodContexts)
        .filter(value => value.methodRunIndex === myRunIndex-1 || value.methodRunIndex === myRunIndex+1)
        .map(value => new MethodDetails(value, this._methodDetails.classStatistics))
        .forEach(value => {
            if (value.methodContext.methodRunIndex > myRunIndex) {
                this._nextMethod = value
            } else {
                this._prevMethod = value
            }
        })
})
```

```html
<div if.bind="_prevMethod">
    <span class="secondary sr1">Previous failed method</span>
     <a route-href="route: method; params.bind: {methodId: _prevMethod.methodContext.contextValues.id}" class="sr1">${_prevMethod.identifier}</a>
</div>
```

Reply:
Thanks for your hints. So best practice is to calculate complex stuff only via `activate` method and cache the results?

Reply:
For Aurelia:
- Prevent calling functions in the view, use observable properties instead.

General:
- Do it when needed
- Do it efficient
- Do it once

Reply:
I updated `methods.ts`. But I added some filters for test methods and failed tests. Finally based on a sort list to run index I can get the previous and the next method.